### DISCUSSION TITLE
Question about Options/Preferences/Audio System

### DISCUSSION THREAD

**Discussion by dfkettle** [hydrogen-music/hydrogen#1343] [Hydrogen on Windows]
I don't know if this is expected behaviour or another issue with the Windows version of Hydrogen. When I go to Options/Preferences/Audio System, I don't see any audio devices listed (see screen capture). The sound is getting routed to my default audio device, a USB audio interface, but sometimes I want to send it to a virtual audio cable (I'm using EffectsRack for this). In other applications (Pure Data, Audacity, VCV Rack, etc.) I see the virtual audio cables listed under devices. But in Hydrogen, I don't see any devices (I have two real devices, the audio ports on the motherboard and an external USB interface, which is the default system device). Even if Hydrogen doesn't recognize or support the virtual audio cables, shouldn't I see at least the two real devices on my system and be able to choose which one I want to use? 

If this is a bug in the Windows version, I'll open an issue on Github. Can somebody advise me on whether or not it's a bug?

![Hydrogen](https://user-images.githubusercontent.com/8409615/128514658-ac42fe3a-bf8a-416d-8ced-5cfa9ff210bc.png)

**Comment by cme** [comment]
Yeah, there's currently no way to select a device -- noted here: #611. 

I'm sure I remember there being some way to assign apps to sound devices in Windows as a workaround?

Selecting devices in general could probably be improved a lot, I think the only interface for it at the moment is specifying a textual name. We should refactor this so that the audio driver interfaces can provide a list of devices to the GUI to populate a menu with.

**Reply by dfkettle** [reply]
> Yeah, there's currently no way to select a device -- noted here: #611.
> 

That's from May 2018. I'm guessing Windows isn't very high priority with the support team. Does anyone even support the Windows version?

> I'm sure I remember there being some way to assign apps to sound devices in Windows as a workaround?
> 

I'm not sure what you mean by that.

> Selecting devices in general could probably be improved a lot, I think the only interface for it at the moment is specifying a textual name. We should refactor this so that the audio driver interfaces can provide a list of devices to the GUI to populate a menu with.

You mean I can just type in a name? I'll try that. Here's what I see in Audacity, for example. Since Audacity is open source too, and runs on Windows, MacOS and Linux, maybe you could re-use some of their code.

Thanks!

![Hydrogen](https://user-images.githubusercontent.com/8409615/128521026-da874551-a0ca-42b7-bfe8-a54af081c7fd.png)

**Reply by dfkettle** [reply]
Update: If I select "PortAudio" for Audio System, it won't let me type in anything for Device. If I select "Auto", it allows me to type in a name, but it seems to ignore whatever I type.

**Reply by cme** [reply]
> > Yeah, there's currently no way to select a device -- noted here: #611.
> That's from May 2018. I'm guessing Windows isn't very high priority with the support team. Does anyone even support the Windows version?

As a spare-time open source project, Hydrogen runs entirely on volunteer energy and enthusiasm. There isn't exactly a "support team" of any sort, only a community-minded spirit and a desire to help people where we can. :) Windows is a problem in general since the developers tend not to be Windows users, so don't often have access to Windows machines to develop and test, but if an issue is actually causing a problem for people, that's a good motivator for someone to try and actually fix it.

Case in point: #1344 :)

> > I'm sure I remember there being some way to assign apps to sound devices in Windows as a workaround?
> I'm not sure what you mean by that.

Sorry, I mean: does Windows have the ability to map applications to specific output devices? I'm sure I remember seeing a mixing desk type GUI somewhere. However, again I'm not generally a Windows user so my memory is foggy!

> > Selecting devices in general could probably be improved a lot, I think the only interface for it at the moment is specifying a textual name. We should refactor this so that the audio driver interfaces can provide a list of devices to the GUI to populate a menu with.
> 
> You mean I can just type in a name? 

For drivers which support naming a device, yes. PortAudio doesn't do so yet. But a better system in now in progress :)

**Reply by dfkettle** [reply]
> As a spare-time open source project, Hydrogen runs entirely on volunteer energy and enthusiasm. There isn't exactly a "support team" of any sort, only a community-minded spirit and a desire to help people where we can. :) Windows is a problem in general since the developers tend not to be Windows users, so don't often have access to Windows machines to develop and test, but if an issue is actually causing a problem for people, that's a good motivator for someone to try and actually fix it.

I'd be happy to help test the Windows version. I have computers running Windows 7, 8 and 10, as well as Ubuntu Studio in a dual-boot configuration with my Windows 7 computer. I don't know that my programming skills would be sufficient to make the modifications myself, but I'm guessing you use a cross-compiler of some kind to create the Windows version, so I wouldn't have to compile it myself.

> Sorry, I mean: does Windows have the ability to map applications to specific output devices? I'm sure I remember seeing a mixing desk type GUI somewhere. However, again I'm not generally a Windows user so my memory is foggy!

Not that I know of, unless it's some 3rd party application and not part of Windows itself. If there is some app I could install for that purpose, I'd be interested in trying it.

**Reply by cme** [reply]
> I'd be happy to help test the Windows version. I have computers running Windows 7, 8 and 10, as well as Ubuntu Studio in a dual-boot configuration with my Windows 7 computer.

Wonderful! If we have a release candidate with this included in the next couple of days, would you be able to give it a once-over before the weekend? :)

(I've also added the same feature for macOS's CoreAudio and added a selection menu for Linux's ALSA driver)

> I don't know that my programming skills would be sufficient to make the modifications myself, but I'm guessing you use a cross-compiler of some kind to create the Windows version, so I wouldn't have to compile it myself.

We've previously used MXE cross-compilation environment to build the Windows releases on Linux, but now we have Appveyor continuous integration building Windows binaries in a well-defined repeatable manner.

**Reply by dfkettle** [reply]
Sure, send me instructions on how to download it when it's ready.

**Reply by cme** [reply]
Here are Appveyor builds:

 - https://ci.appveyor.com/api/buildjobs/wbs1uexnyamcmil9/artifacts/Hydrogen-1.1.0-win32.exe -- 32-bit Windows installer
 - https://ci.appveyor.com/api/buildjobs/11ndicu995436967/artifacts/Hydrogen-1.1.0-win64.exe -- 64-bit Windows installer

I guess one concern I have is that, due to the fact that PortAudio exposes the same device multiple times for different underlying host APIs, it can appear a bit redundant and confusing, but I don't think it's excessively so.

**Reply by dfkettle** [reply]
Didn't have much success, I'm afraid. On my Windows 7 & 10 computers, it displayed a list of devices (repeating some several times, as you said), but it just seemed to ignore my selection. Audio would be sent to the default audio device, no matter what I selected. Strangely enough, it seemed to work on my Windows 8 computer, although the program would freeze for a couple of seconds in the middle of playback, at random intervals. It seems a little unstable in general, compared to the current production release. Maybe because you used a different compiler?

I don't really understand why the devices would be repeated in the list. That doesn't happen with any other audio apps I have, like Audacity, Pure Data, Reaper, etc. Since Hydrogen just uses PortAudio, can't you filter out all the other entries?

**Reply by cme** [reply]
> Didn't have much success, I'm afraid. On my Windows 7 & 10 computers, it displayed a list of devices (repeating some several times, as you said), but it just seemed to ignore my selection. Audio would be sent to the default audio device, no matter what I selected. 

Hmm, interesting. The audio driver needs a restart to change device, and there *should* be a notification that says this when you hit 'OK' on the dialog box and offers to restart (or use the 'Apply and restart output' button next to the device list), but it looks like this is bypassed when you only change the device, not the driver. I'll fix that. In the meantime, if you can try hitting that 'Apply and restart' button that should change the device.

> Strangely enough, it seemed to work on my Windows 8 computer, although the program would freeze for a couple of seconds in the middle of playback, at random intervals. It seems a little unstable in general, compared to the current production release. Maybe because you used a different compiler?

The compiler shouldn't have any effect here, but since the rest of the environment is different it could easily be something there or in the libraries. In particular, selecting a non-default PortAudio device / API might cause problems. If it's repeatable, if you could enable logging and check the log file (or reproduce here) that would be useful!

![image](https://user-images.githubusercontent.com/768466/129453521-f854a0a1-a781-4a0d-bf2a-78b92a7141ae.png)

> I don't really understand why the devices would be repeated in the list. That doesn't happen with any other audio apps I have, like Audacity, Pure Data, Reaper, etc. Since Hydrogen just uses PortAudio, can't you filter out all the other entries?

Just in case you're curious for the actual answer to the "why" question -- PortAudio is a wrapper and supports multiple underlying host APIs. Windows has WASAPI, WMDKS, DirectSound etc. and when asked to enumerate devices, PortAudio reports devices multiple times for each API it's accessible by.

I suspect what most software does is pick the default API and report devices for that. It reduces the duplication but has the possibility of potentially missing some devices.

(Sadly there doesn't seem to be any sensible way of telling that the same device appearing via different APIs is really the same device to allow that to be deduplicated.)

Thanks for your help!

**Reply by dfkettle** [reply]
> Hmm, interesting. The audio driver needs a restart to change device, and there _should_ be a notification that says this when you hit 'OK' on the dialog box and offers to restart (or use the 'Apply and restart output' button next to the device list), but it looks like this is bypassed when you only change the device, not the driver. I'll fix that. In the meantime, if you can try hitting that 'Apply and restart' button that should change the device.

I did click on 'Apply and restart output'. Made no difference.

> The compiler shouldn't have any effect here, but since the rest of the environment is different it could easily be something there or in the libraries. In particular, selecting a non-default PortAudio device / API might cause problems. If it's repeatable, if you could enable logging and check the log file (or reproduce here) that would be useful!

I'll try that and post the log file later.

> I suspect what most software does is pick the default API and report devices for that. It reduces the duplication but has the possibility of potentially missing some devices.

Actually, no. Most audio software on Windows lets you choose which API to use. For example, Audacity (on my Windows 7 computer) lets me choose MME, Windows DirectSound or Windows WASAPI. Pure Data only offers a choice of MMIO or ASIO via PortAudio, but it does let me choose. Reaper offers a choice of WDM Kernel Streaming, DirectSound, WaveOut (I'm not sure what that is), ASIO, "Dummy Audio" or WASAPI. And they all let me choose different devices for input and output, but in the case of Hydrogen, I guess that isn't really necessary. 

Audacity:

![Audacity](https://user-images.githubusercontent.com/8409615/129455814-1ea5386f-8485-44e3-9bf0-459c846f1c8a.png)

Reaper:

![Reaper](https://user-images.githubusercontent.com/8409615/129456001-7ac8d840-4989-4832-b723-c8f0b04d8d82.png)

**Reply by dfkettle** [reply]
OK, there seems to be an error reported when I click on 'Apply and restart output':

(E) PortAudioDriver::connect Found but can't open device 'Line (3- Steinberg UR22)': Invalid number of channels[0m
(E) PortAudioDriver::connect Can't use device 'Line (3- Steinberg UR22)', using default stream[0m

But the error message doesn't really make any sense. My default device is a Steinberg UR22, which has 2 input and 2 output channels. And if I open Hydrogen without setting a device in Preferences (in other words, just let it default), this is the device that gets used, and it works fine! But if I try to select a device (any device, including the default device), I get the above error.

Attached is the complete log file. See lines 55-56 for the error message.

[hydrogen.log](https://github.com/hydrogen-music/hydrogen/files/6987035/hydrogen.log)

**Reply by cme** [reply]
> Most audio software on Windows lets you choose which API to use. For example, Audacity (on my Windows 7 computer) lets me choose MME, Windows DirectSound or Windows WASAPI.

Aaah, this seems so obviously the right thing to do now it's been pointed out :) I was lulled into cul-de-sac by macOS having only a single host API that made me fail to consider it should just be an extra level of hierarchy. 

> But the error message doesn't really make any sense. My default device is a Steinberg UR22, which has 2 input and 2 output channels. And if I open Hydrogen without setting a device in Preferences (in other words, just let it default), this is the device that gets used, and it works fine! But if I try to select a device (any device, including the default device), I get the above error.

Great, thanks! I have a suspicion for what the issue is here. I note that "Line (3- Steinberg UR22)" doesn't explicitly say "line *out*"; is there also a line *in* that shows up with the same name? If so (and PortAudio is treating them as separate devices with the same name) then Hydrogen could mistakenly pick up the input device when trying to open the output stream.

Thanks again for your help here, this has been fantastically useful!

**Reply by dfkettle** [reply]
> Great, thanks! I have a suspicion for what the issue is here. I note that "Line (3- Steinberg UR22)" doesn't explicitly say "line _out_"; is there also a line _in_ that shows up with the same name? If so (and PortAudio is treating them as separate devices with the same name) then Hydrogen could mistakenly pick up the input device when trying to open the output stream.

There is a line in as well, but I'm not sure if would have the same name. Is it possible to filter out the input devices? That would probably reduce the duplication in the list, too. Some devices are showing up 3 times in the list for some reason. I don't think Hydrogen would mistake an input device for an output device, but the only way to be sure would be to look at the source code. It doesn't seem to matter which of the duplicate entries I select; they all seem to give an error.

**Reply by cme** [reply]
Would you mind giving this build a try, please?

https://ci.appveyor.com/api/buildjobs/2sjv13vkq91q9k09/artifacts/Hydrogen-1.1.0-win64.exe

(though note that it's a debug build so will pop open a console window to display the logging info)

**Reply by dfkettle** [reply]
I'm able to select the output device now, but the sound is very distorted for any output device except "Primary Sound Driver" or "Line (Steinberg UR22)". Note that my default device, a Steinberg UR22, is showing up 4 times in the list: 3 times as "Line (3 - Steinberg UR22" and once as "Line (Steinberg UR22". Other devices are also showing up 4 times, with one description appearing 3 times and a different description appearing once. I really think you need to find some way to filter out the duplicates or it will just confuse users. Sorting the list might be helpful too, as well as making it easier to remove duplicate entries. I've attached 3 screen prints showing the complete list of devices. (The third screen print overlaps with the second one.)

![DeviceList1](https://user-images.githubusercontent.com/8409615/129742467-bbe7c3ab-c30d-41d0-af69-0a79977d7286.png)
![DeviceList2](https://user-images.githubusercontent.com/8409615/129742470-7bd16342-b1e9-4553-810f-d3b6a154d406.png)
![DeviceList3](https://user-images.githubusercontent.com/8409615/129742471-af4aa6be-a29e-4c5b-a887-24c2fba434a2.png)

As for the distortion, it sounds as if there's a delay with feedback being applied, even though I'm not applying any effects at all. I've attached two sound files, 'test-good.txt' and 'test-bad.txt' (renamed because this forum only allows certain file types to be uploaded), to demonstrate. You'll have to rename them back to *.wav to listen to them (hopefully). I've also attached the pattern I used to test it (renamed as "Pattern.txt").
 and the log file.

[hydrogen.log](https://github.com/hydrogen-music/hydrogen/files/7000534/hydrogen.log)
[Pattern.txt](https://github.com/hydrogen-music/hydrogen/files/7000549/Pattern.txt)
[test-bad.txt](https://github.com/hydrogen-music/hydrogen/files/7000561/test-bad.txt)
[test-good.txt](https://github.com/hydrogen-music/hydrogen/files/7000563/test-good.txt)

**Reply by cme** [reply]
> Other devices are also showing up 4 times, with one description appearing 3 times and a different description appearing once. I really think you need to find some way to filter out the duplicates or it will just confuse users.

Sorry, there's a small initial state bug there. The intention is that the device menu should now show only the device entries for the selected HostAPI. Without a HostAPI selected, it'll still show all of them. I have managed to verify this works as expected for me on a Windows VM (though slowly).

> As for the distortion, it sounds as if there's a delay with feedback being applied, even though I'm not applying any effects at all.

Thaks for including the audio, that really helped. It looks like opening the non-default device isn't handling sample rate correctly, so I'll review that. Thanks again!

ETA: On closer inspection it's not the sample rate, but a timing or buffering issue that you're seeing, and it's consistent with one that I've managed to reproduce on a faster VM, and fix. 

(The `Pa_OpenDefaultStream` call that's used (and was always the one used previously) when no device is specified, uses the device's 'high' default latency value, so that it's safe as a fallback. This is not documented in the API documentation. Using `Pa_OpenStream` and specifying a minimum latency, the latency is device-specific and may not be achievable, particularly for things like USB devices, in which case PortAudio will miss deadlines and emit silence for some buffer periods. The fix is to use the device default 'high' latency just like `Pa_OpenDefaultStream`)

If you feel like trying this, a release build with these changes is at:
  https://ci.appveyor.com/api/buildjobs/deahbc1d2l5hou47/artifacts/Hydrogen-1.1.0-win64.exe

**Comment by dfkettle** [comment]
Hi! Sorry for the delay in responding, I didn't realize you'd created another build. It's great! The device list is a lot cleaner now, and the sound  isn't distorted. I've only tested this on my Windows 7 computer, but I'll test it under Windows 8 & 10 as well, later today.

![Hydrogen](https://user-images.githubusercontent.com/8409615/130244316-e4ec291a-d55a-4fbe-b672-61af55fc7d40.png)

Thanks!

**Comment by dfkettle** [comment]
I tested it under Windows 8 & 10 and it works as well.

Would you happen to know the status of another issue I reported (https://github.com/hydrogen-music/hydrogen/discussions/1147)? I'm not getting the error message I was before, but I'm not sure if it's actually been fixed. I have some plug-ins installed for other programs I'm using, but I'm not sure if any of them are LADSPA plug-ins (the only kind of plug-in supported by Hydrogen, is that right)?). If I can find some LADSPA plug-ins compiled for Windows, I can test that as well.

**Reply by cme** [reply]
> I tested it under Windows 8 & 10 and it works as well.
> 
> Would you happen to know the status of another issue I reported (#1147)? I'm not getting the error message I was before, but I'm not sure if it's actually been fixed. I have some plug-ins installed for other programs I'm using, but I'm not sure if any of them are LADSPA plug-ins (the only kind of plug-in supported by Hydrogen, is that right)?). If I can find some LADSPA plug-ins compiled for Windows, I can test that as well.

LADSPA plugins *should* work in that Windows build, which ought to be fairly representative of what we release as 1.1, so if you feel like giving it a go, that would be really useful! They didn't work on the 1.0.2 release builds because the MXE cross compilation environment doesn't explicitly support it, but in retrospect all that's needed is a header file so I *could* have supported in the binaries, if only I'd realised that.

Thanks again!

**Reply by dfkettle** [reply]
> LADSPA plugins _should_ work in that Windows build, which ought to be fairly representative of what we release as 1.1, so if you feel like giving it a go, that would be really useful! They didn't work on the 1.0.2 release builds because the MXE cross compilation environment doesn't explicitly support it, but in retrospect all that's needed is a header file so I _could_ have supported in the binaries, if only I'd realised that.

I'm afraid I haven't had any luck with installing plugins, but it isn't necessarily anything wrong in Hydrogen itself, I just don't know what I'm doing. After extensive googling, I found one post stating that you had to define an environment variable named LADSPA_PATH and set it to the directory containing the plugins (I don't think this is mentioned in the official Hydrogen manual). So I did that, but nothing appears in the plugins list, even after closing and re-opening Hydrogen. However, I'm not sure if any of the plugins I have installed are LADSPA plugins. So I have a silly question: how do I know if a plugin is a LADSPA plugin? They all have 'dll' extensions, but I don't know what kind of plugins they are (I was hoping at least one or two might be LADSPA). 

I tried to find some LADSPA plugins for Windows to download, but all the plugins I found were compiled for Linux. Some had instructions for compiling under Windows, but I was hoping to find some that were already compiled for Windows. Any idea what I'm doing wrong, if anything, or where I might find some LADSPA plugins compiled for Windows?

![Hydrogen](https://user-images.githubusercontent.com/8409615/130336164-08e10933-cf83-47ea-a030-e87d4c576179.png)

**Reply by theGreatWhiteShark** [reply]
> I found one post stating that you had to define an environment variable named LADSPA_PATH and set it to the directory containing the plugins (I don't think this is mentioned in the official Hydrogen manual).

This is indeed the biggest (known) missing subject in the manual. While rewriting the whole thing I didn't came across a comprehensive guide to LADSPA on Windows but only a number of not that recent posts. Since I have no Windows machine available I decided to not write anything before writing up something wrong. It would be great if you could help to add this missing piece but right now I don't have the resources to do so. In a month or two would be better. For the time being: some FLOSS applications state that they are shipped with a number of LADSPA plugins, like Audacity or Radium.  Maybe these can be used.

**Reply by dfkettle** [reply]
> This is indeed the biggest (known) missing subject in the manual. While rewriting the whole thing I didn't came across a comprehensive guide to LADSPA on Windows but only a number of not that recent posts. Since I have no Windows machine available I decided to not write anything before writing up something wrong. It would be great if you could help to add this missing piece but right now I don't have the resources to do so. In a month or two would be better. For the time being: some FLOSS applications state that they are shipped with a number of LADSPA plugins, like Audacity or Radium. Maybe these can be used.

Audacity is supposed to support LADSPA plugins, but I don't think any LADSPA plugins are installed by default. But I'm not sure, since they all have 'dll' extensions under Windows, so I can't tell what kind of plugins they are. So I downloaded Radium (which I'd never heard of) and set the environment variable LADSPA_PATH to point to the directory in Radium containing the LADSPA plugins (\<Radium installation directory\>\bin\ladspa). I can now see the plugins listed in Hydrogen, but they don't seem to do anything. I tried selecting several different effects, but they seemed to make no difference to the sound (I clicked the [BYP] button to turn the effect on and off). Is there something else I have to do?

If I can get it working, I'll document the steps I followed so you can update the manual. Some of the steps may be slightly different, depending on whether you're on Windows 7, 8, or 10, but I'll try to cover them.

BTW, some of the LADSPA effects were stereo, but most were mono. I tried stereo effects as well as mono. Should a mono effect be applied to both left and right channels, or will it just be ignored or cause an error?

![Hydrogen1](https://user-images.githubusercontent.com/8409615/130361980-98b5e289-b9b4-4490-ada3-82ae63df029e.png)
![Hydrogen2](https://user-images.githubusercontent.com/8409615/130361981-9fa56e6f-5645-4ad3-badd-7f55ee0ac35f.png)
![Hydrogen3](https://user-images.githubusercontent.com/8409615/130361982-a7734b88-5123-46de-99e5-fc1bd4b1e427.png)

**Reply by theGreatWhiteShark** [reply]
> BTW, some of the LADSPA effects were stereo, but most were mono. I tried stereo effects as well as mono. Should a mono effect be applied to both left and right channels, or will it just be ignored or cause an error?

The general/expected behavior would be that the mono effect is opened twice by the application for both the left and right channel and the program just shows a single window to control the parameters of both of them. But I wouldn't be surprise if Hydrogen handles it not as expected.

Personally, I think the plugins support in the mixer is close to embarrassing. One can only use aux sends instead of adding effects to single tracks. While having aux sends is in general a good idea, the total number of 4 is extremely limiting. (I need an EQ for each instrument to cut out all the resonance and clean up improper recording before I can use a kit for more than a rough demo). The overall UI is super odd. Why faders for the parameters with a meter that is not moving at all but just represents the current value? It's not adding any information but might be misleading. In the long run we might rework the whole thing but it's not on the priority list right now.

**Reply by cme** [reply]
> After extensive googling, I found one post stating that you had to define an environment variable named LADSPA_PATH and set it to the directory containing the plugins (I don't think this is mentioned in the official Hydrogen manual).

We really ought to add some sort of GUI for this. Dependencies on environment variables are bad especially on Windows and macOS where users aren't even expected to know what those are.

**Reply by dfkettle** [reply]
> We really ought to add some sort of GUI for this. Dependencies on environment variables are bad especially on Windows and macOS where users aren't even expected to know what those are.

@theGreatWhiteShark @cme Agreed. But in the meantime, here are instructions for Windows 7, 8 and 10. BTW, Audacity is still 32-bit, so any plugins included with it will be 32-bit. So they won't work in Hydrogen (unless you downloaded the 32-bit version of Hydrogen, I guess). That's why I had to download Radium. Would it be possible to bundle the effects (since they're all free, although they may not all be open-source) and include them in a sub-directory of Hydrogen (say \<Hydrogen install directory\>\LADSPA)? Then you wouldn't have to set any environment variables, since Hydrogen would just assume they're in that directory.

Windows 7:

1. Click on "Start" (button in lower left corner).
2. Click on "Control Panel".
3. Click on "System and Security".
4. Click on "System".
5. Click on "Advanced system settings".
6. Click on "Environment Variables".
7. Under either "User variables" (for yourself only) or "System variables"
   (for all users), click on "New...".
8. For "Variable name", enter "LADSPA_PATH".
9. For "Variable value", enter the full path to directory containing plugins.
10. In "Environemt variables" window, click on "OK".
11. In "Advanced system settings" window, click on "OK".
12. In "System" window, click on "OK".
13. Close "System and Security" window.
14. If Hydrogen is open, close and re-open.

Windows 8:

1. Right-click on the "Start" (button in lower left corner).
2. Go to step 2 above.

Windows 10:

1. Right-click on the "Start" (button in lower left corner).
2. Click on "Settings".
3. Go to step 5 above.

**Reply by dfkettle** [reply]
Correction: Seems Audacity is 64-bit, so I'm not sure why the plugins included with it weren't recognized. Anyway, it seems the same plugins are included with Radium, and they were recognized.

**Reply by theGreatWhiteShark** [reply]
Also the [LMMS doc](https://docs.lmms.io/user-manual/6-built-in-effects/6.17-ladspa-plugins) suggests it is bundled and shipped with the most essential LADSPA plugins.

**Comment by cme** [comment]
Funny thing, I've never actually tried using LADSPA effects, but... yeah, I don't hear any difference. A quick look at the code and... nothing ever seems to actually fill the input buffers of the LADSPA filters. They're connected up in `audioEngine_setupLadspaFX` with the same buffers as input and output, which doesn't seem right either,  but it's been this way at least since 2014. I'll have a dig into the history to try and make sense of this. @theGreatWhiteShark you've used LADSPA effects, right?

**Reply by dfkettle** [reply]
So it's not just a Windows issue?

**Reply by cme** [reply]
Well, I could easily be misunderstanding how it works, but it *looks* to me like it is broken\* and has been so for a long time.

Any synthesis type plugin that creates an output but doesn't require an input (are there any of these?) would work, because the output of the plugin is added into the audio output. But this seems wrong, doesn't it? I thought all plugins were supposed to function as filters.

I wouldn't be surprised if I'm just entirely misunderstanding it.

**Reply by dfkettle** [reply]
In the VST world, VSTi's (short for Vst instruments) don't require any audio input (they receive MIDI data instead). I don't know about LADSPA, but these are supposed to be effects, so I think, by definition, they require input.

**Reply by theGreatWhiteShark** [reply]
> Any synthesis type plugin that creates an output but doesn't require an input (are there any of these?) would work, because the output of the plugin is added into the audio output. But this seems wrong, doesn't it? I thought all plugins were supposed to function as filters.

There are a number of plugins that are agnostic to their input, like https://tomscii.sig7.se/tap-plugins/ladspa/pinknoise.html . But I think the general concept is that they receive an input buffer, do some processing, and hand it over to the next element in the processing chain. In principle it should be possible to do some wave synthesize on the basis of an empty track with e.g. http://quitte.de/dsp/caps.html#Sin . But nowadays you would use a more convenient graphical LV2 plugin.

> @theGreatWhiteShark you've used LADSPA effects, right?

I use/used them a lot (but mostly outside of Hydrogen). I just tried at 4355e93e and it works fine.

**Reply by cme** [reply]
So, it turns out that I just forgot how these are wired up (in both code terms, and in GUI terms).

In the code, the filter input buffer is written by the sampler on a per-note basis, before the mixer.

Correspondingly, in the GUI, each instrument's send level for each filter has to be set (defaulting to 0). The GUI isn't very helpful and I don't think it's clear in the manual. But once I remembered that, it all seems to work fine.

[Video, Hydrogen on Windows with a LADSPA filter](https://youtu.be/XbZpRCb_9T8)

The latency is horrible but that's partially due to running in a VM, and partially due to #1352 . The low audio qualiy and background noise is due it being acoustic from my laptop's speakers to microphone.

**Reply by dfkettle** [reply]
> Correspondingly, in the GUI, each instrument's send level for each filter has to be set (defaulting to 0). The GUI isn't very helpful and I don't think it's clear in the manual. But once I remembered that, it all seems to work fine.

OK, now I understand. You have to turn up one of the four tiny knobs as well as the knob labelled "RETURN". I'm not sure if there's a latency problem, but I hear the effect now.
